### ISSUE TITLE
Certificate authority is always expected

### ISSUE THREAD

**Issue by garethahealy** [openshift/jenkins-client-plugin#94]
The plugin either expects that the CA is configured via "Cluster Configuration" (i.e.: via Manage Jenkins page) or, it defaults to the mounted ca.crt under /var/run.

This is fine, if you are using the plugin to talk to the same cluster you are running in or you want to explicitly set the CA. But in some instances, customers will include the signer of the CA into the Jenkins slaves, thus the certificate-authority is not needed.

For example, if my jenkins is running on cluster1, but trying to talk to cluster2, with the current plugin, it would use the following command:

`oc  get projects --server=https://cluster2.corp:8443 --namespace=dev --certificate-authority=/var/run/secrets/kubernetes.io/serviceaccount/ca.crt --token=XXXXX `

Which will fail with a 509x cert issue.

If i manually run the same command, without the certificate-authority specified, the command succeeds.

Is it possible to provide a flag, either by Cluster Config or Env, which can turn off the use of certificate-authority?

FYI: I am a RedHatter (part of UK Consulting @gahealy) on-site with a customer.

**Comment by bparees** [comment]
Setting the CA as part of the cluster configuration doesn't seem that unreasonable to me (presumably you have to do other cluster configuration anyway), though the ability to use the default CA baked into a slave image makes sense.

**Comment by garethahealy** [comment]
Currently, yes, i need to create a cluster configuration, but if i could disable the plugin adding the certificate-authority onto the command line, i could just do the below, and it would work, because the cluster2 is already trusted by the jenkins slave:

`stages {
    stage ('Do something') {
        steps {
            script {
                openshift.withCluster("https://cluster2.corp:8443") {
                    openshift.withProject("test") {
                        //do something within project
                    }   
                }
            }
        }
    }
}`

But because there is no cluster configuration, and the above is running in cluster1, the plugin try's to connect to cluster2 with cluster1s certificate-authority.

**Comment by gabemontero** [comment]
@garethahealy - remind me, if the CA is already baked in, does Java store it in a well known location of the file system, or does some env var need to be set?

If so, then when a URL is the value passed into `withCluster`, we could just determine if the CA is baked in.  If so, avoid setting the `oc` option.  

I'm also wondering if defaulting to the pod cert when an arbitrary URL is passed in makes sense.  It is sort of a last ditch effort / assumption that the mounted cert would work.  Perhaps we should default to skip tls verify, and if a user wants full SSL, create an appropriate jenkins cluster config, etc.

@bparees @jim-minter @coreydaley any thoughts there?

**Comment by bparees** [comment]
> Perhaps we should default to skip tls verify

we should never do that :)

As to whether we should be defaulting to the mounted pod cert(as we currently do), or just hoping/assuming the system certs in the image will suffice, I can see arguments for both.  Allowing the user to explicitly indicate they want to use the image's system certs (ie they do not want the plugin to add a `--certificate-authority` flag to the oc command) seems reasonable to me, I think that's what's being asked for here.

**Comment by gabemontero** [comment]
Yeah it is just what to do when they do *NOT* specify that they want to use the image's certs.  As you said, and as to what lead to me originally bringing it up, there are arguments both ways.

**Comment by bparees** [comment]
> Yeah it is just what to do when they do NOT specify that they want to use the image's certs.

since we've already established that if they don't specify anything, they get the mounted pod certs, i don't think we can change that (it could break people in unexpected ways), so i think the only path forward is to allow users to explicitly request that we do not use the mounted pod certs (so that oc falls back to the system certs)

**Comment by bparees** [comment]
btw i don't think java is involved, this is all about what ca certs the oc client binary is going to use (either the ones provided as an arg to oc by the plugin, or the system defaults)

**Comment by gabemontero** [comment]
On Mon, Jan 1, 2018 at 7:08 PM, Ben Parees <notifications@github.com> wrote:

> btw i don't think java is involved, this is all about what ca certs the oc
> client binary is going to use (either the ones provided as an arg to oc by
> the plugin, or the system defaults)
>
my interpretation was they were using the typical java tooling to bake in
the cert, and then let oc leverage it, but sure, that could be a
misinterpretation

**Comment by garethahealy** [comment]
As @bparees has said, the certs are added at OS level via _update-ca-trust_

**Comment by gabemontero** [comment]
ok thanks - I believe I've connected the dots

so, how to expose this option to the user via the client plugin surgary dsl syntax (that has been giving me a fair amount of heartburn this week ;-) ) ?

Any opinions folks?

Both withCredentials (formerly doAs) and withCluster have credentialIds ... do we just add a `ignoreMountedCerts` flag on those?

**Comment by bparees** [comment]
> Both withCredentials (formerly doAs) and withCluster have credentialIds ... do we just add a ignoreMountedCerts flag on those?

sounds right to me.  I was thinking "useSystemCerts" but we don't have to debate the name extensively.

**Comment by muff1nman** [comment]
Another Red Hatter here at a customer. This is also a problem for us and the use case we have is for an external Jenkins talking to an openshift cluster and the Jenkins host OS trust store (`/etc/pki/ca-trust/source/anchors/`) needs to be used for tls purposes. The default (`/var/run/secrets/kubernetes.io/serviceaccount/ca.crt`) makes no sense when jenkins is not running on openshift. A flag might not be needed if a reliable check can be done to see if jenkins is running on openshift or not, but we could work with a `useSystemCerts` flag as well.

**Comment by gabemontero** [comment]
For now opened https://trello.com/c/6AlVfZ2X/1511-2-allow-use-system-certs-ignore-mounted-certs-for-client-plugin-pipelineintegration

Before following the process and closing this out in favor of the card, I'll double check with @bparees - do we still want to consider this as an enhancement, or instead pivot and consider it a "shortcoming bordering on problem that needs to be fixed"
